pipelines/common/src/main/java/com/google/fhir/analytics/FetchUtil.java

```diff
@@ -0,0 +1,199 @@
+      return (Resource) resource;
+    } catch (Exception e) {
+      log.error(
+          String.format(
+              "Failed fetching FHIR %s resource with Id %s: %s", resourceType, resourceId, e));
+      return null;
+    }
+  }
+
+  public Resource fetchFhirResource(String resourceUrl) {
+    // Parse resourceUrl
+    String[] sepUrl = resourceUrl.split("/");
+    String resourceId = sepUrl[sepUrl.length - 1];
+    String resourceType = sepUrl[sepUrl.length - 2];
+    return fetchFhirResource(resourceType, resourceId);
+  }
+
+  public IGenericClient getSourceClient() {
+    return getSourceClient(false);
+  }
+
+  public IGenericClient getSourceClient(boolean enableRequestLogging) {
+    fhirContext.getRestfulClientFactory().setSocketTimeout(200 * 1000);
+
+    IGenericClient client = fhirContext.getRestfulClientFactory().newGenericClient(fhirUrl);
+    if (authInterceptor != null) {
+      // TODO add logic to inspect/refresh token if we get 401/403 responses.
+      client.registerInterceptor(authInterceptor);
+    }
+
+    if (enableRequestLogging) {
+      LoggingInterceptor loggingInterceptor = new LoggingInterceptor();
+      loggingInterceptor.setLogger(log);
+      loggingInterceptor.setLogRequestSummary(true);
+      client.registerInterceptor(loggingInterceptor);
+    }
+
+    return client;
+  }
+
+  public String getSourceFhirUrl() {
+    return fhirUrl;
+  }
+
+  /**
+   * A simple class for fetching OAuth tokens through the client_credential grant flow and injecting
+   * them to FHIR API calls. Instances of this class are thread-safe and should be shared between
+   * threads for better performance and to reduce the load on the token endpoint.
+   */
+  private static class ClientCredentialsAuthInterceptor extends BearerTokenAuthInterceptor {
+
+    private static final int TOKEN_REFRESH_LEEWAY_IN_SECONDS = 10;
+
+    private final String tokenEndpoint;
+    private final String clientId;
+    private final String clientSecret;
+    private TokenResponse tokenResponse;
+    private Instant nextRefresh;
+
+    ClientCredentialsAuthInterceptor(String tokenEndpoint, String clientId, String clientSecret) {
+      Preconditions.checkNotNull(tokenEndpoint);
+      Preconditions.checkNotNull(clientId);
+      Preconditions.checkNotNull(clientSecret);
+      this.tokenEndpoint = tokenEndpoint;
+      this.clientId = clientId;
+      this.clientSecret = clientSecret;
+    }
+
+    @Override
+    public synchronized String getToken() {
+      if (nextRefresh == null || Instant.now().isAfter(nextRefresh)) {
+        try {
+          log.debug("Fetching a new OAuth token; old refresh: {}", nextRefresh);
+          tokenResponse = requestAccessToken();
+          log.debug("Fetched token is: {}", tokenResponse.getAccessToken());
+        } catch (IOException e) {
+          log.error("Failed fetching a new token through client_credentials grant_type", e);
```

Comment:
Should we propagate back this error? Returning `null` will end in adding this null in the `Bearer` token and might further fail as `Unauthenticated User`?

Reply:
Done (I cannot throw a checked exception because of `@Override`).